packages/peregrine/lib/talons/CheckoutPage/__tests__/useCheckoutPage.spec.js

```diff
@@ -0,0 +1,209 @@
+    }
+]);
+const placeOrder = jest.fn();
+const placeOrderMutationResult = jest.fn().mockReturnValue([
+    placeOrder,
+    {
+        error: null,
+        loading: false,
+        data: null
+    }
+]);
+
+/**
+ * Helpers
+ */
+
+const Component = props => {
+    const talonProps = useCheckoutPage(props);
+
+    return <i talonProps={talonProps} />;
+};
+
+const getTalonProps = props => {
+    const tree = createTestInstance(<Component {...props} />);
+    const { root } = tree;
+    const { talonProps } = root.findByType('i').props;
+
+    const update = newProps => {
+        act(() => {
+            tree.update(<Component {...{ ...props, ...newProps }} />);
+        });
+
+        return root.findByType('i').props.talonProps;
+    };
+
+    return { talonProps, tree, update };
+};
+
+/**
+ * beforeAll
+ */
+
+beforeAll(() => {
+    useLazyQuery.mockImplementation(query => {
+        if (query === getCheckoutDetailsQuery) {
+            return getCheckoutDetailsQueryResult();
+        } else if (query === getOrderDetailsQuery) {
+            return getOrderDetailsQueryResult();
+        } else {
+            return [jest.fn(), {}];
+        }
+    });
+
+    useMutation.mockImplementation(mutation => {
+        if (mutation === createCartMutation) {
+            return createCartMutationResult();
+        } else if (mutation === placeOrderMutation) {
+            return placeOrderMutationResult();
+        } else {
+            return [jest.fn(), {}];
+        }
+    });
+
+    useApolloClient.mockReturnValue(client);
+});
+
+/**
+ * Tests
+ */
+
+test('Should return correct shape', () => {
+    const { talonProps } = getTalonProps(props);
+
+    expect(talonProps).toMatchSnapshot();
+});
+
+test('Should get checkout details if cartId changes and order has not been placed yet', () => {
+    useCartContext.mockReturnValueOnce([
+        { cartId: '123' },
+        { createCart: jest.fn(), removeCart: jest.fn() }
+    ]);
+
+    const { update } = getTalonProps(props);
+
+    expect(getCheckoutDetails.mock.calls[0][0]).toMatchObject({
+        variables: { cartId: '123' }
+    });
+    expect(getCheckoutDetails.mock.calls.length).toBe(1);
+
+    useCartContext.mockReturnValueOnce([
+        { cartId: 'abc' },
+        { createCart: jest.fn(), removeCart: jest.fn() }
+    ]);
+
+    update();
+
+    expect(getCheckoutDetails.mock.calls[1][0]).toMatchObject({
+        variables: { cartId: 'abc' }
+    });
+    expect(getCheckoutDetails.mock.calls.length).toBe(2);
+
+    useCartContext.mockReturnValueOnce([
+        { cartId: 'xyz' },
+        { createCart: jest.fn(), removeCart: jest.fn() }
+    ]);
+    placeOrderMutationResult.mockReturnValueOnce([
+        placeOrder,
+        {
+            error: null,
+            loading: false,
+            data: {
+                placeOrder: { order: { order_number: '000000' } }
+            }
+        }
+    ]);
+
+    update();
+
+    expect(getCheckoutDetails.mock.calls.length).toBe(2);
```

Comment:
Consider splitting this out into multiple tests so that you can follow the Arrange, Act, Assert pattern (AAA).

More details here: https://freecontent.manning.com/making-better-unit-tests-part-1-the-aaa-pattern/.

Reply:
Not relevant anymore. Updated the tests to replicate the new loading check. 0efb6563523d2a24a9a8d58b3f98ec6633c436ff